fix(app): ensure run timer is set if you connect during a delay

celsasser opened the pull request:
# DRAFT 
Before testing I would like to make sure the approach is acceptable. Will check with @mcous and follow up one way or the other.

<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

# Changelog

<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests

<!--
Describe any requests for your reviewers here.
-->

# Risk assessment

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

Commit:
fix(app): Run timer stays at 00:00:00 if you reconnect in the middle of a delay:
Starting over and taking the the option behind curtain number two in the PR to revise the way pause and runtime operate. Actually it's really well spelled out in the old PR so going to drop a variation of it in here:
- Remove the runTime state entry because it represents "now, if running", and we can calculate that from startTime in a selector / in the UI component
- Remove remoteTimeCompensation state entry and associated machinery, because with time syncing, this has been made unnecessary
- Remove the TICK_RUN_TIME action and associated machinery, because why would we need to do this in state?
- Total run time is now driven by:
   - Hasn't run yet: nothing, it's 0
   - When running / paused: Date.now() less session.startTime
   - When stopped / errored / finished: session.statusInfo.changedAt (because that's when the status changed to stopped, etc.)
- Current paused time is driven by:
   - If not paused: nothing, it's 0
   - If paused: Date.now() less session.startTime + session.statusInfo.changedAt
   - If robot is paused, user knows clearly that it's paused and for how long it's been paused, because UI clearly communicates a ticking time labeled "Paused For"
- tore out `remoteTimeCompensation`
- tore out `TICK_RUN_TIME` (UI no longer updates. Need to fix)
- some crumby styling

Commit:
Going back and forth with strategy for support getRunTime and getPausedTime and trying to replication. I am settling down with what might be a unpalatable use of selectors and will run it by the community before testing.
- revise timestamp dependent selectors to return a method that takes a timestamp. Motivations for doing so are commented within.

mcous commented on app/src/redux/robot/selectors.ts:
```diff
@@ -212,49 +213,76 @@ export const getSessionError: (state: State) => string | null = createSelector(
+ */
+export const getRunSecondsAt: (
+  state: State
+) => (now?: number) => number = createSelector(
```
I think this might be a lot easier to grok if we just ditched `createSelector` entirely. I agree with your assessment that there aren't really performance concerns, and this codebase has a fairly unhappy history with selector factories.

```ts
export const getRunSeconds(state: State, now: number) => number = {
  statusInfo = getSessionStatusInfo(state)
  isRunning = getIsRunning(state)
  isDone = getIsDone(state)

  // ...
}
```

celsasser replied:
Ah, I was not understanding that we were taking the `createSelector` duties out of the selector.  Sure.

mcous commented on app/src/redux/robot/selectors.ts:
```diff
@@ -212,49 +213,76 @@ export const getSessionError: (state: State) => string | null = createSelector(
+    isDone: boolean
+  ) => (now: number = Date.now()): number => {
+    if (isRunning) {
+      // TODO: to the reviewers, at this point, I think logically `startTimeMs` should not be null.
```
I agree, but I also think `if (isRunning && startTimeMs !== null) {` here is fairly harmless if it lets us avoid non-null TS assertion

celsasser replied:
Sure. The problem with that that I did not like is any value returned without factoring in `startTimeMs` seems wrong?

celsasser replied:
The rationale behind similar choices within were motivated by the same concern.

mcous commented on app/src/pages/Run/RunPanel/RunTimer.tsx:
```diff
@@ -1,23 +1,53 @@
+  /**
+   * Using a a timer to tick at a 1 second interval to update the run time and pause durations.
+   */
+  React.useEffect(() => {
```
Intervals are weirder with `useEffect` than they should be. Luckily, we've "solved the problem" (read: copy-pasted [the solution from Dan Abramov](https://overreacted.io/making-setinterval-declarative-with-react-hooks/)) already:

```ts
import { useInterval } from '@opentrons/components'

// ...

useInterval(() => setNow(Date.now()), 1000)
```

mcous commented on app/src/pages/Run/RunPanel/RunTimer.tsx:
```diff
@@ -1,23 +1,53 @@
+export function RunTimer(): JSX.Element {
+  const [now, setNow] = React.useState(Date.now())
+  const startTimeMs = useSelector(getStartTimeMs)
+  const getRunSeconds = useSelector(getRunSecondsAt)(now)
```
This will bust the cache of our selector every single time, which I think gives more ammo to the "ditch `createSelector`" idea

mcous commented on app/src/pages/Run/RunPanel/RunTimer.tsx:
```diff
@@ -1,23 +1,53 @@
 
   return (
     <div>
+      <div className={styles.pause_time}>
```
Just to make sure we get to it: this should only be rendered if paused

celsasser replied:
Hmmm, I was working off of feedback in https://github.com/Opentrons/opentrons/pull/7841:

> ...we should update the run timer UI when the protocol is paused, doing something like:
> 
> Make the currently big "Run Time: 01:02:03" smaller, calling it "Total Run Time" and putting it just above "Start Time"
> Where the big runner timer used to be, put "Paused For: 01:02:03"
> 

I don't think I'm clear on the design.

mcous replied:
This is what I had in mind, does this make more sense? (Excuse the delay in switching over to the paused time: the component needs access to the status to do that better)

![2021-06-08 07 00 42](https://user-images.githubusercontent.com/2963448/121174392-d9e6e680-c827-11eb-9edb-646724a42094.gif)

mcous commented on app/src/pages/Run/RunPanel/styles.css:
```diff
@@ -18,6 +18,18 @@
 }
 
 .run_time {
+  padding: var(--run-panel-padding);
```
This isn't obvious when you're working in old components like `RunTimer`, but we no longer use CSS Modules for new components. If possible, could we switch over to styled-components and/or our [basic UI atom components](https://s3-us-west-2.amazonaws.com/opentrons-components/edge/index.html?path=/story/library-atoms-icon--icon)?

celsasser replied:
Hoping this can wait for another rainy day?

mcous reviewed (commented):
Haven't 👁️'d the UI yet, but I think this is on the right track. Assorted comments below

Commit:
Addressed feedback but not doing such a wonderful job. Pretty crappy. Going to work on some flow -> TS tickets.
- addressing some of the feedback on the draft.

mcous commented on components/src/styles/typography.css:
```diff
@@ -4,6 +4,7 @@
   --c-font-light: white;
   --c-font-disabled: #9c9c9c;
   --fs-huge: 3rem;
+  --fs-big: 2.25rem;
```
Generally, I think we shouldn't generally be messing with our typography scale without input from UI (also I don't think we need it for this PR)

mcous commented on app/src/pages/Run/RunPanel/RunTimer.tsx:
```diff
@@ -1,23 +1,48 @@
+  const [now, setNow] = React.useState(Date.now())
+  const state = useStore().getState()
+  const startTimeMs = useSelector(getStartTimeMs)
+  const runSeconds = getRunSeconds(state, now)
+  const pausedSeconds = getPausedSeconds(state, now)
```
Even if a given selector function may not be a [reselect](https://github.com/reduxjs/reselect) `createSelector` selector, you should still use `useSelector` over `useStore`:

```suggestion
  const runSeconds = useSelector(state => getRunSeconds(state, now))
  const pausedSeconds = useSelector(state => getPausedSeconds(state, now))
```

mcous commented on app/src/redux/robot/api-client/client.js:
```diff
@@ -426,30 +425,9 @@ export function client(dispatch) {
-    }
-  }
-
-  function clearRunTimerInterval() {
```
There are still calls to `clearRunTimerInterval` that need to be removed

Commit:
Merge branch 'edge' into fix-run-timer-revision

Commit:
Merging changes from `edge`

Commit:
Addressing some of the feedback. Still need to:
- replace css components with basic UI atom components (see PR feedback)
- manually and thoroughly test all scenarios
- unit test changes

Commit:
- have made a mess

celsasser commented:
please retest this
